crypto/ml_kem/ml_kem.c

```diff
@@ -0,0 +1,1732 @@
+/*
+ * Copyright 2024 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+/* Copyright (c) 2024, Google Inc. */
+
+#ifndef OPENSSL_NO_ML_KEM
+
+# include <assert.h>
+# include <internal/common.h>
+# include <internal/constant_time.h>
+# include <internal/sha3.h>
+# include <crypto/ml_kem.h>
+# include <openssl/rand.h>
+
+/*
+ * Structure of keys
+ */
+typedef struct ossl_ml_kem_scalar_st {
+    /* On every function entry and exit, 0 <= c[i] < ML_KEM_PRIME. */
+    uint16_t c[ML_KEM_DEGREE];
+} scalar;
+
+/* General form of public and private key storage */
+#define DECLARE_ML_KEM_KEYDATA(name, rank, private_sz) \
+    struct ossl_ml_kem_##name##_st { \
+        /* Public vector |t| */ \
+        scalar tbuf[(rank)]; \
+        /* Pre-computed matrix |m| */ \
+        scalar mbuf[(rank)*(rank)] \
+        /* optional private key data */ \
+        private_sz \
+    }
+
+/* Declare variant-specific public and private storage */
+#define DECLARE_ML_KEM_VARIANT_KEYDATA(bits) \
+    DECLARE_ML_KEM_KEYDATA(bits##_puballoc, ML_KEM_##bits##_RANK,;); \
+    DECLARE_ML_KEM_KEYDATA(bits##_prvalloc, ML_KEM_##bits##_RANK,;\
+        scalar sbuf[ML_KEM_##bits##_RANK]; \
+        uint8_t zbuf[ML_KEM_RANDOM_BYTES];)
+DECLARE_ML_KEM_VARIANT_KEYDATA(512);
+DECLARE_ML_KEM_VARIANT_KEYDATA(768);
+DECLARE_ML_KEM_VARIANT_KEYDATA(1024);
+#undef DECLARE_ML_KEM_VARIANT_KEYDATA
+#undef DECLARE_ML_KEM_KEYDATA
+
+typedef const ML_KEM_VINFO *vinfo_t;
+typedef __owur
+int (*cbd_t)(scalar *out, uint8_t in[ML_KEM_RANDOM_BYTES + 1],
```

Comment:
cbd_t = CBD_FUNC or CBD_CALLBACK

Reply:
This is private to just that file. I prefer to keep it short.